File: python-sample-app/python_sample_app/hitcounter.py
from constructs import Construct
from aws_cdk import (
    aws_lambda as _lambda,
    aws_dynamodb as ddb
)


class HitCounter(Construct):

    @property
    def handler(self):
        return self._handler

    def __init__(self, scope: "Construct", id: 
        str, downstream: _lambda.IFunction, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        table = ddb.Table(
            self, 'Hits',
            # https://beabetterdev.com/2022/02/07/dynamodb-partition-key-vs-sort-key/
            # TLDR: partition key is like a primary key for dynamodb and is required
            partition_key={'name': 'path', 'type': ddb.AttributeType.STRING}
        )

        self._handler = _lambda.Function(
            self, 'HitCountHandler',
            runtime=_lambda.Runtime.PYTHON_3_7,
            handler='hitcount.handler',
            code=_lambda.Code.from_asset('lambda'),
            # table_name and function_name are `late-bound values`;
            #  properties that are only resolved when stack is deployed
            environment={
                'HITS_TABLE_NAME': table.table_name,
                'DOWNSTREAM_FUNCTION_NAME': downstream.function_name
            }
        )
        
        # must give lambda read/write permissions to our db table
        table.grant_read_write_data(self.handler)
        # must give this lambda permissions to invoke downstream lambda
        downstream.grant_invoke(self.handler)

File: typescript-sample-app/lib/hitcounter.ts
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

export interface HitCounterProps {
    downstream: lambda.IFunction;
}

export class HitCounter extends Construct {
    
    /** allows accessing the counter function */
    // similar to @property of python class
    public readonly handler: lambda.Function;

    /** allows accessing the hitcounter dynamo table */
    public readonly table: dynamodb.Table;

    constructor(scope: Construct, id: string, props: HitCounterProps) {
        super(scope, id);
        
        this.table = new dynamodb.Table(this, 'Hits', {
            partitionKey: {
                name: 'path', type: dynamodb.AttributeType.STRING
            },
            removalPolicy: cdk.RemovalPolicy.DESTROY
        });

        this.handler = new lambda.Function(
            this, 'HitCounterHandler', {
                runtime: lambda.Runtime.NODEJS_14_X,
                code: lambda.Code.fromAsset('lambda'),
                handler: 'hitcounter.handler',
                environment: {
                    DOWNSTREAM_FUNCTION_NAME: props.downstream.functionName,
                    HITS_TABLE_NAME: this.table.tableName
                }
            }
        );

        // must give lambda read/write permissions to db table
        this.table.grantReadWriteData(this.handler);

        // must give hitcounter lambda permissions to invoke downstream lambda
        props.downstream.grantInvoke(this.handler);
    }
}